feat: Add Copy command parsing in QueryMessage and basic psql e2e test

Vizerai opened the pull request.

Vizerai committed:
Adding Copy command parsing in QueryMessage and basic e2e test.

Vizerai committed:
Update

Vizerai committed:
Update

Vizerai committed:
Update

tinaspark commented on src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/ControlMessage.java:
```diff
@@ -49,35 +50,46 @@ public ControlMessage(ConnectionHandler connection) throws IOException {
+          return new CopyDoneMessage(connection);
+        case CopyDataMessage.IDENTIFIER:
+          return new CopyDataMessage(connection);
+        default:
```
should we handle CopyFail here as well?

Vizerai replied:
Done.

tinaspark commented on src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/ControlMessage.java:
```diff
@@ -49,35 +50,46 @@ public ControlMessage(ConnectionHandler connection) throws IOException {
+          return new SyncMessage(connection);
+        case TerminateMessage.IDENTIFIER:
+          return new TerminateMessage(connection);
+        case CopyDoneMessage.IDENTIFIER:
```
are we allowed to receive CopyDone/CopyData/CopyFail if we're not in COPY_IN mode?

Vizerai replied:
No, it should not.

olavloite commented on src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDoneMessage.java:
```diff
@@ -51,13 +51,15 @@ protected void sendPayload() throws Exception {
         // Spanner returned an error when trying to commit the batch of mutations.
         mw.writeMutationsToErrorFile();
         mw.closeErrorFile();
+        this.connection.setStatus(ConnectionStatus.IDLE);
```
Sorry, I didn't notice this before, but don't we need to return something like a `ReadyResponse` after this? Or will the exception that is thrown ensure that the `ConnectionHandler` returns a response?

Vizerai replied:
The ConnectionHandler will return a ready response of IDLE on handling an error.

olavloite commented on src/main/java/com/google/cloud/spanner/pgadapter/wireprotocol/CopyDoneMessage.java:
```diff
@@ -51,13 +51,15 @@ protected void sendPayload() throws Exception {
       mw.closeErrorFile();
     }
     new ReadyResponse(this.outputStream, ReadyResponse.Status.IDLE).send();
+    this.connection.setStatus(ConnectionStatus.IDLE);
```
I think that this call to `setStatus(..)` and also the above one are causing a build error, because `ConnectionStatus` is not imported in this file. Would you mind taking a look at that?

Vizerai replied:
Fixed now.  Originally this PR was supposed to require changes from https://github.com/GoogleCloudPlatform/pgadapter/pull/40

olavloite reviewed (commented):
The build fails because of a compile error, I think because `ConnectionStatus` is not imported by `CopyDoneMessage`.

Vizerai committed:
Update: refactor some of the code and fix tests.

Vizerai commented:
The e2e failures seem to be due to a change on the backend.  There a new pg_catalog entries returned from the information schema.

olavloite committed:
Merge branch 'postgresql-dialect' into copyparse

olavloite commented:
The current e2e test failure should be fixed once https://github.com/GoogleCloudPlatform/pgadapter/pull/40 has been merged (it's caused by the "?" that is fixed in #40)

tinaspark reviewed (commented):
LGTM once the test failures are fixed by #40

Vizerai committed:
Merge branch 'postgresql-dialect' into copyparse

Vizerai committed:
Merge branch 'postgresql-dialect' of https://github.com/GoogleCloudPlatform/pgadapter into copyparse

Vizerai committed:
Merge branch 'copyparse' of https://github.com/GoogleCloudPlatform/pgadapter into copyparse

Vizerai committed:
Update

olavloite committed:
fix: do not return Ready response for Copy messages

Copy messages should not respond with a Ready response, only with its
own specific Copy responses. Failure to do so can cause the calling
client to think that the operation has already finished, which again can
cause the entire copy operation to fail if the connection is closed
directly after receiving the Ready response.

Also the current Copy operation should not execute a rollback if
something fails, as it is operating in auto-commit mode. Trying to
rollback when in auto-commit mode will cause an error.

olavloite committed:
test: update test values to avoid duplicates

olavloite committed:
fix: remove non-existing rows from expected file

olavloite committed:
fix: add missing COPY to response

olavloite committed:
fix: copy comes before the select

olavloite committed:
fix: COPY is on stdin, the expected result has an extra empty line

olavloite commented:
@Vizerai I was wondering why the e2e-tests were failing, so I had a look at this PR and ended up doing a couple of small changes. It might be good if you go through those to check that you are OK with them before merging.

Vizerai commented:
> @Vizerai I was wondering why the e2e-tests were failing, so I had a look at this PR and ended up doing a couple of small changes. It might be good if you go through those to check that you are OK with them before merging.

The changes look good.  Thanks for the update.

Vizerai committed:
Adding /*GSQL*/ prefix to information schema query to get spanner types
instead of PG types.

tinaspark commented on src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java:
```diff
@@ -169,11 +169,11 @@ private void queryInformationSchema() throws SQLException {
     PreparedStatement statement =
         this.connection.prepareStatement(
-            "SELECT "
+            "/*GSQL*/SELECT "
```
why do we need to issue this as a GoogleSQL query?

Vizerai replied:
The backend no longer returns these as GoogleSQL types for this query. It instead returns the PG types now. We need the GoogleSQL types to fill out the mutations.

This was just recently changed:
https://buganizer.corp.google.com/issues/198312860

tinaspark replied:
I see. Can you add a comment?

// The mutation API requires GoogleSQL type names instead of PostgreSQL type names for the table column types
// Issue the information_schema table query as a GoogleSQL query so that it will return GoogleSQL type names.

Vizerai replied:
Done.

Vizerai committed:
Adding comment about /*GSQL*/ information_schema query.

Vizerai committed:
Update expected test results.

Vizerai committed:
Update

Vizerai committed:
Update
